File: apps/backend/generator-service/src/lib/db/generatordb.py
# IMPORT THE SQALCHEMY LIBRARY's CREATE_ENGINE METHOD
from sqlalchemy import text
from lib.logger import initLogger
from lib.db.postgresdb import PostgresDB
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from domain.entity.base import Base
from domain.entity.job import Job
from domain.entity.result import Result
import os


# PYTHON FUNCTION TO CONNECT TO THE POSTGRESQL DATABASE AND
# RETURN THE SQLACHEMY ENGINE OBJECT
# this class is purposely as exposed as possible as the DB can always change (eg. new tables etc.)
# it's as simple as possible to expose the engine and session to the repository layer.

# DB
class GeneratorDB(PostgresDB):
    def __init__(self):
        self.engine = None
        self.session = None
        
    def connect_db(self):
        logger = initLogger("generatorDB.connect_db")
        logger.info("Connecting to DB...")
        try:
            logger.info("Creating engine...")
            engine = self.get_connection() 
            self.engine = engine

            logger.info("Creating session...")
            Session = sessionmaker(bind=engine)
            self.session = Session()
        except Exception as e:
            logger.error(f"Failed to connect to DB: {e}")

    def get_connection(self):
        # DEFINE THE DATABASE CREDENTIALS
        user = os.getenv("GENERATOR_DB_USER")
        password = os.getenv("GENERATOR_DB_PASSWORD")
        host = os.getenv("GENERATOR_DB_HOST")
        port = os.getenv("GENERATOR_DB_PORT")
        database = os.getenv("GENERATOR_DB_NAME")

        engine = create_engine(
            url="postgresql://{0}:{1}@{2}:{3}/{4}".format(
                user, password, host, port, database
            ),
            echo=True
        )
        return engine
    
    def set_all_tables(self):
        logger = initLogger("generatorDB.set_all_tables")
        logger.info("Setting tables...")
        Base.metadata.create_all(self.engine)

    def reset_all_tables(self):
        logger = initLogger("generatorDB.reset_all_tables")
        logger.info("Resetting tables...")

        try:
            with self.engine.connect() as conn:
                conn.execute(text("DROP TABLE IF EXISTS results CASCADE;"))
                conn.execute(text("DROP TABLE IF EXISTS jobs CASCADE;"))
                conn.commit()  # commit the transaction if needed

            Base.metadata.create_all(self.engine)
        except Exception as e:
            logger.error(f"Failed to reset tables: {e}")
            raise

    
# create a singleton
generatorDB = GeneratorDB()

File: apps/backend/api/app/lib/db/generatordb.go
package db

import (
	"fmt"
	"os"

	"github.com/jolneetanky/dateideas/apps/backend/api/app/domain/entity"
	"github.com/jolneetanky/dateideas/apps/backend/api/app/lib/logger"
	"gorm.io/driver/postgres"
	"gorm.io/gorm"
)

type GeneratorDB struct {
	GormDB *gorm.DB
}

func InitGeneratorDB() GeneratorDB {
	return GeneratorDB{}
}

// Implement methods
// Receiver needs to be a pointer so we can modify `pg.DB`.
func (pg *GeneratorDB) ConnectDB() error {
	dsn := fmt.Sprintf(
		"host=%s user=%s password=%s dbname=%s port=%s sslmode=disable TimeZone=Asia/Shanghai",
		os.Getenv("GENERATOR_DB_HOST"),
		os.Getenv("GENERATOR_DB_USER"),
		os.Getenv("GENERATOR_DB_PASSWORD"),
		os.Getenv("GENERATOR_DB_NAME"),
		os.Getenv("GENERATOR_DB_PORT"),
	)

	// Open DB connection
	logger.Info("Connecting to DB...")
	db, err := gorm.Open(postgres.Open(dsn), &gorm.Config{})

	if err != nil {
		logger.Error(fmt.Sprintf(
			"Failed to connect to DB. Error: %s",
			err.Error(),
		))
		return err
	}

	(*pg).GormDB = db

	logger.Info("Successfully connected to DB")

	return nil
}

func (pg GeneratorDB) ResetAllTables() error {
	logger.Info("Resetting all tables...")

	// Drop the tables if they exist
	err := pg.GormDB.Migrator().DropTable(&entity.Result{}, &entity.Job{})
	if err != nil {
		logger.Error(fmt.Errorf("failed to drop tables: %w", err))
		return err
	}

	// Recreate the tables
	err = pg.GormDB.AutoMigrate(&entity.Job{}, &entity.Result{})
	if err != nil {
		logger.Error(fmt.Errorf("failed to migrate tables: %w", err))
		return err
	}

	logger.Info("All tables reset successfully.")
	return nil
}
